File: python/ch06_agent/agent_ch06/core/tools.py
"""工具注册表。

这是什么：工具的注册、准备、调用和定义管理
Java 类比：@Service class ToolRegistry + CommandHandler 接口
为什么需要：集中管理所有可用工具，提供统一的调用接口和参数校验

Java 对照：`ToolRegistry` 类似按命令名保存 Handler 的注册表；`prepare` 是
进入 Service 前的 JSON/schema 校验，`invoke` 才真正执行副作用。
"""

import json
import re
from collections.abc import Callable, Mapping
from dataclasses import dataclass
from types import MappingProxyType
from typing import Any

from .messages import ToolCall
from .model import OpenAIToolSchema

EffectClass = str  # 工具的副作用分类，用于权限决策


@dataclass(frozen=True, slots=True)
class ToolContext:
    """程序提供给工具的受控运行环境。

    这是什么：传递给工具执行器的上下文对象
    Java 类比：record ToolContext(String workspace, String identity)
    为什么需要：封装工具执行所需的环境信息（工作目录、身份），避免工具自行猜测

    工具不能自己猜工作目录，而是由 AgentRunner 明确传入，类似 Java Service
    接收一个包含租户、用户和工作目录的上下文对象。
    """

    workspace: str  # 工具执行时使用的受控工作目录。
    identity: str  # 发起调用的用户或 Agent 身份。


@dataclass(frozen=True, slots=True)
class ToolResult:
    """工具的统一返回值。

    这是什么：工具执行后的结果对象，成功或失败都用此类型
    Java 类比：record ToolResult(String content, boolean isError, String errorCode)
    为什么需要：统一工具的返回格式，让模型能看到错误并自行调整策略

    工具失败也返回 ToolResult，而不是直接让整个 Agent 崩溃。
    这样模型能看到错误，并有机会换一种做法。
    """

    content: str  # 给模型看的结果文本。
    is_error: bool  # True 表示执行失败，但仍然是合法的工具结果。
    error_code: str | None = None  # 机器可读错误码，例如 shell_timeout。


def tool_success(content: str) -> ToolResult:
    """创建成功结果；错误码必须保持为空。

    这是什么：工具成功结果的工厂方法
    Java 类比：public static ToolResult success(String content)
    为什么需要：提供类型安全的成功结果构造，确保错误码为空
    """
    return ToolResult(content, False)


def tool_error(error_code: str, message: str) -> ToolResult:
    """创建失败结果，让模型看到可理解的错误，而不是 Python 堆栈。"""
    if not error_code.strip():
        raise ValueError("工具错误码不能为空")
    return ToolResult(f"工具执行错误 [{error_code}]: {message}", True, error_code)


def copy_tool_result(result: ToolResult) -> ToolResult:
    """复制工具结果，切断 Hook 或 handler 持有的原对象引用。"""
    if not isinstance(result, ToolResult):
        raise TypeError("只能复制 ToolResult")
    return ToolResult(result.content, result.is_error, result.error_code)


ToolHandler = Callable[[Mapping[str, Any], ToolContext], ToolResult]
"""工具处理器签名，作用类似 Java 的 `BiFunction<Arguments, Context, Result>`。"""


@dataclass(frozen=True, slots=True)
class ToolDefinition:
    """一个完整工具定义：名称、说明、参数格式、副作用类型和执行函数。"""

    name: str  # 注册表中的唯一名称。
    description: str  # 发送给模型的工具说明。
    parameters: dict[str, Any]  # JSON Schema 参数约束。
    effect: EffectClass  # 副作用分类，当前为 read/write/execute/external 之一。
    handler: ToolHandler  # 参数校验通过后真正执行的函数。
    validator: Callable[[Mapping[str, Any]], bool] | None = None  # 执行前的严格参数校验器。


@dataclass(frozen=True, slots=True)
class PreparedToolCall:
    """已经完成查找、JSON 解析和参数校验的工具调用。

    如果准备失败，`error` 中已经放好了要返回给模型的错误结果，
    AgentRunner 不需要用异常分支处理未知工具或错误 JSON。
    """

    call: ToolCall  # 模型原始调用，必须保留 ID。
    definition: ToolDefinition | None = None  # 找到的工具定义。
    arguments: Mapping[str, Any] | None = None  # JSON 解析、校验并冻结后的参数。
    error: ToolResult | None = None  # 准备阶段失败时直接回填的错误。


def _freeze_json(value: Any) -> Any:
    """递归冻结 JSON 数据。

    Java 对照：类似把 Map/List 递归转换成 `Map.copyOf` 和 `List.copyOf`。
    `MappingProxyType` 是 Python 标准库提供的只读 Map 视图。
    """
    if isinstance(value, dict):
        return MappingProxyType({key: _freeze_json(item) for key, item in value.items()})
    if isinstance(value, list):
        return tuple(_freeze_json(item) for item in value)
    return value


def copy_prepared_tool_call(
    prepared: PreparedToolCall,
    *,
    definition: ToolDefinition | None = None,
) -> PreparedToolCall:
    """复制并冻结 prepared call，同时按需保留可信工具定义的对象身份。"""
    if not isinstance(prepared, PreparedToolCall):
        raise TypeError("只能复制 PreparedToolCall")
    copied_arguments = (
        None if prepared.arguments is None else _freeze_json(dict(prepared.arguments))
    )
    copied_error = None if prepared.error is None else copy_tool_result(prepared.error)
    return PreparedToolCall(
        ToolCall(prepared.call.id, prepared.call.name, prepared.call.arguments),
        definition if definition is not None else prepared.definition,
        copied_arguments,
        copied_error,
    )


class ToolRegistry:
    """集中管理所有模型可以调用的工具。"""

    def __init__(
        self, definitions: dict[str, ToolDefinition] | None = None, mutable: bool = True
    ) -> None:
        # 复制一份字典，避免外部继续修改传入的 definitions 影响本注册表。
        # Java 对照：类似构造器里 new HashMap<>(definitions)。
        self._definitions = dict(definitions or {})
        # True 表示启动装配阶段允许 register；快照会改成 False。
        self._mutable = mutable

    @property
    def names(self) -> tuple[str, ...]:
        """返回已注册工具名的不可变快照，便于日志和测试查看。"""
        return tuple(self._definitions)

    def register(self, definition: ToolDefinition) -> None:
        """注册工具，并在启动阶段尽早发现重名或非法名称。"""
        if not self._mutable:
            raise ValueError("工具注册表快照不可修改")
        if not re.fullmatch(r"[A-Za-z0-9_]+", definition.name):
            raise ValueError(f"工具名称不合法: {definition.name}")
        if not definition.description.strip():
            raise ValueError("工具描述不能为空")
        if definition.name in self._definitions:
            raise ValueError(f"工具已经注册过: {definition.name}")
        self._definitions[definition.name] = definition

    def snapshot(self) -> "ToolRegistry":
        """生成只读快照。

        模型这一轮看到哪些工具，就必须只能执行这些工具，不能在请求过程中偷偷变化。
        """
        return ToolRegistry(self._definitions, mutable=False)

    def openai_tools(self) -> tuple[OpenAIToolSchema, ...]:
        """把 Python 工具定义转换成模型能理解的 JSON Schema。"""
        return tuple(
            OpenAIToolSchema(d.name, d.description, d.parameters)
            for d in self._definitions.values()
        )

    def prepare(self, call: ToolCall) -> PreparedToolCall:
        """准备工具调用，但绝不执行真实副作用。

        顺序是：按名字找工具 -> 把 arguments 字符串解析成 JSON -> 校验必须是对象
        -> 校验 shell 只允许 command 字段。任何错误都转换成 ToolResult。
        """
        definition = self._definitions.get(call.name)
        if definition is None:
            return PreparedToolCall(
                call, error=tool_error("unknown_tool", f"找不到工具: {call.name}")
            )
        try:
            raw = json.loads(call.arguments)
        except json.JSONDecodeError:
            return PreparedToolCall(
                call,
                definition=definition,
                error=tool_error("invalid_json", "工具参数必须是合法 JSON"),
            )
        if not isinstance(raw, dict):
            return PreparedToolCall(
                call,
                definition=definition,
                error=tool_error("invalid_arguments", "工具参数必须是 JSON 对象"),
            )
        if definition.validator is not None and not definition.validator(raw):
            return PreparedToolCall(
                call,
                definition=definition,
                error=tool_error("invalid_arguments", "工具参数没有通过格式校验"),
            )
        # prepare 成功后立刻冻结参数。后面的 Hook、审批器和 handler 共享这份快照，
        # 外部代码无法在异步等待期间偷偷修改参数。
        return copy_prepared_tool_call(PreparedToolCall(call, definition=definition, arguments=raw))

    def invoke(self, prepared: PreparedToolCall, context: ToolContext) -> ToolResult:
        """真正执行已经准备好的工具调用。

        这是副作用发生的位置。handler 抛出的异常会在这里统一转换，
        不把 Python 堆栈或敏感信息直接发给模型。
        """
        if prepared.error is not None:
            return prepared.error
        if prepared.definition is None or prepared.arguments is None:
            raise ValueError("准备好的工具调用不完整")
        try:
            result = prepared.definition.handler(prepared.arguments, context)
            if not isinstance(result, ToolResult):
                return tool_error("invalid_tool_result", "工具处理函数返回了无效结果")
            if result.is_error and not result.error_code:
                return tool_error("invalid_tool_result", "工具处理函数返回了无效结果")
            if not result.is_error and result.error_code is not None:
                return tool_error("invalid_tool_result", "工具处理函数返回了无效结果")
            return result
        # handler 是可插拔的外部代码，所有异常都要在注册表边界转换成稳定结果。
        except Exception:  # noqa: BLE001
            return tool_error("tool_execution_error", "工具执行失败")

File: typescript/chapters/ch18/src/core/tools.ts
import { z } from "zod";

import type { ToolCall } from "./messages.js";
import { toolCall } from "./messages.js";
import type { OpenAIToolSchema } from "./model.js";

// effect 是权限策略和工具列表可见性的最小事实来源，不依赖工具名称猜测。
export type EffectClass = "read" | "write" | "execute" | "external";
// concurrency 是 Dispatcher 判断工具能否转后台执行的显式契约，缺省为 inline。
export type ConcurrencyClass = "inline" | "background_eligible";

// ToolContext 是每次工具调用的不可变执行边界；P18 由 provider 在 handler 前补入
// taskId/claimToken/worktreeName，并把 workspace 解析成当前 claim 的 Worktree 路径。
export interface ToolContext {
  // 实际执行目录：未绑定 claim 时是主 workspace，绑定后是受管 Worktree。
  readonly workspace: string;
  // 执行者身份，用于 claim owner 校验；由 Runner 固定，handler 不可修改。
  readonly identity: string;
  // 幂等键与显式认领 token：自动认领/Subagent 用同一 token 跨回复恢复 claim。
  readonly idempotencyKey?: string;
  readonly taskId?: string;
  readonly claimToken?: string;
  readonly worktreeName?: string;
  // 同一 assistant 回复内的短时 claim 关联，WorktreeRuntime 用它回查 claim token。
  readonly executionScope?: object;
}

export interface ToolResult {
  // content 回填模型；失败必须额外携带稳定 errorCode。
  readonly content: string;
  readonly isError: boolean;
  readonly errorCode?: string;
}

// 成功和失败都返回不可变 ToolResult；错误必须携带稳定错误码。
export function toolSuccess(content: string): ToolResult {
  return Object.freeze({ content, isError: false });
}

export function toolError(errorCode: string, message: string): ToolResult {
  if (errorCode.trim().length === 0) {
    throw new Error("tool error code must not be empty");
  }
  // 失败结果必须提供稳定 errorCode，便于模型、Hook 和存储层分类处理。
  return Object.freeze({
    content: `Error [${errorCode}]: ${message}`,
    isError: true,
    errorCode,
  });
}

// 任何进入 canonical history 的工具结果都深拷贝，防止 handler 内部可变对象污染会话。
export function copyToolResult(result: ToolResult): ToolResult {
  if (!isToolResult(result)) {
    throw new Error("tool result must satisfy the ToolResult contract");
  }
  if (!result.isError) {
    return toolSuccess(result.content);
  }
  const errorCode = result.errorCode;
  if (errorCode === undefined) {
    throw new Error("error tool result requires an errorCode");
  }
  return Object.freeze({ content: result.content, isError: true, errorCode });
}

export interface ToolDefinition<Input> {
  // handler 接收 provider 已解析的上下文，不自行决定工作目录或身份。
  readonly name: string;
  readonly description: string;
  readonly inputSchema: z.ZodType<Input>;
  readonly effect: EffectClass;
  readonly concurrency?: ConcurrencyClass;
  readonly handler: (input: Input, context: ToolContext) => Promise<ToolResult> | ToolResult;
}

// StoredToolDefinition 隐藏泛型 Input，统一把 schema 校验放在 invoke 边界。
export interface StoredToolDefinition {
  readonly name: string;
  readonly description: string;
  readonly inputSchema: z.ZodType<unknown>;
  readonly effect: EffectClass;
  readonly concurrency: ConcurrencyClass;
  readonly invoke: (input: unknown, context: ToolContext) => Promise<ToolResult>;
}

export interface PreparedToolCall {
  // 成功时具有 definition/arguments，失败时具有可直接回填的 error。
  readonly call: ToolCall;
  readonly definition?: StoredToolDefinition;
  readonly arguments?: unknown;
  readonly error?: ToolResult;
}

export function freezePreparedToolCall(
  call: ToolCall,
  definition: StoredToolDefinition,
  argumentsValue: unknown,
): PreparedToolCall {
  // Hook 改写后重建冻结快照，审批和执行始终读取同一份参数。
  return Object.freeze({
    call: toolCall(call.id, call.name, call.arguments),
    definition,
    arguments: freezeInput(structuredClone(argumentsValue)),
  });
}

// 工具注册表保存冻结定义，并在分发前按上下文与 schema 建立调用边界。
export class ToolRegistry {
  readonly #definitions: Map<string, StoredToolDefinition>;
  readonly #mutable: boolean;

  constructor(definitions: ReadonlyMap<string, StoredToolDefinition> = new Map(), mutable = true) {
    // 复制 Map 隔离调用方修改；snapshot 通过 mutable=false 禁止注册。
    this.#definitions = new Map(definitions);
    this.#mutable = mutable;
  }

  get names(): readonly string[] {
    // 名称快照按注册顺序返回并冻结。
    return Object.freeze([...this.#definitions.keys()]);
  }

  register<Input>(definition: ToolDefinition<Input>): void {
    // 注册时立即冻结定义；运行期 prepare() 只允许解析、校验和深拷贝输入。
    if (!this.#mutable) {
      throw new Error("tool registry snapshot is immutable");
    }
    if (!/^[A-Za-z0-9_]+$/.test(definition.name)) {
      throw new Error(`invalid tool name: ${definition.name}`);
    }
    if (definition.description.trim().length === 0) {
      throw new Error("tool description must not be empty");
    }
    if (this.#definitions.has(definition.name)) {
      throw new Error(`tool already registered: ${definition.name}`);
    }

    const stored: StoredToolDefinition = Object.freeze({
      name: definition.name,
      description: definition.description,
      inputSchema: definition.inputSchema,
      effect: definition.effect,
      concurrency: definition.concurrency === undefined ? "inline" : definition.concurrency,
      invoke: async (input: unknown, context: ToolContext) =>
        definition.handler(definition.inputSchema.parse(input), context),
    });
    this.#definitions.set(definition.name, stored);
  }

  snapshot(): ToolRegistry {
    // snapshot 用于给子 Agent 等边界提供只读工具视图，不能继续注册新工具。
    return new ToolRegistry(this.#definitions, false);
  }

  subset(names: readonly string[]): ToolRegistry {
    // subset 只暴露显式声明的工具，队友运行时用它限制模型可调用的能力。
    const definitions = new Map<string, StoredToolDefinition>();
    for (const name of names) {
      const definition = this.#definitions.get(name);
      if (definition === undefined) throw new Error(`tool does not exist: ${name}`);
      definitions.set(name, definition);
    }
    return new ToolRegistry(definitions);
  }

  openAITools(): readonly OpenAIToolSchema[] {
    // 模型侧 schema 与运行时 handler 同源，避免平行定义漂移。
    return Object.freeze(
      [...this.#definitions.values()].map((definition) => ({
        type: "function" as const,
        function: {
          name: definition.name,
          description: definition.description,
          parameters: z.toJSONSchema(definition.inputSchema) as Readonly<Record<string, unknown>>,
        },
      })),
    );
  }

  prepare(call: ToolCall): PreparedToolCall {
    // 参数是不可信输入：先解析 JSON，再按 schema 校验，最后冻结完整调用快照。
    const definition = this.#definitions.get(call.name);
    if (definition === undefined) {
      // 工具故障转成可回填消息，循环可继续让模型决定下一步。
      return { call, error: toolError("unknown_tool", `Unknown tool: ${call.name}`) };
    }

    let raw: unknown;
    try {
      raw = JSON.parse(call.arguments);
    } catch {
      return {
        call,
        definition,
        error: toolError("invalid_json", "Tool arguments must be valid JSON"),
      };
    }
    if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
      return {
        call,
        definition,
        error: toolError("invalid_arguments", "Tool arguments must be a JSON object"),
      };
    }

    const parsed = definition.inputSchema.safeParse(raw);
    if (!parsed.success) {
      return {
        call,
        definition,
        error: toolError("invalid_arguments", "Tool arguments failed schema validation"),
      };
    }
    // Pre Hook 只能通过 updatedInput 显式改写，不能就地修改受信任的准备结果。
    return freezePreparedToolCall(call, definition, parsed.data);
  }

  async invoke(prepared: PreparedToolCall, context: ToolContext): Promise<ToolResult> {
    // provider 解析发生在 AgentRunner 调用 invoke 之前。
    if (prepared.error !== undefined) {
      return prepared.error;
    }
    if (prepared.definition === undefined || prepared.arguments === undefined) {
      throw new Error("prepared tool call is incomplete");
    }
    try {
      // handler 异常统一归一为 tool_execution_error，不让内部错误文本直接进入模型上下文。
      const result: unknown = await prepared.definition.invoke(prepared.arguments, context);
      if (!isToolResult(result)) {
        return toolError("invalid_tool_result", "Tool handler returned an invalid result");
      }
      return result;
    } catch {
      return toolError("tool_execution_error", "Tool execution failed");
    }
  }
}

function freezeInput<Input>(value: Input, seen: WeakSet<object> = new WeakSet()): Input {
  if (typeof value !== "object" || value === null) {
    return value;
  }
  if (seen.has(value)) {
    return value;
  }
  seen.add(value);
  for (const nested of Object.values(value)) {
    freezeInput(nested, seen);
  }
  return Object.freeze(value);
}

// handler 返回值同样属于不可信边界，阻止畸形对象污染会话历史。
export function isToolResult(value: unknown): value is ToolResult {
  if (typeof value !== "object" || value === null) {
    return false;
  }
  const content = Reflect.get(value, "content");
  const isError = Reflect.get(value, "isError");
  const errorCode = Reflect.get(value, "errorCode");
  if (typeof content !== "string" || typeof isError !== "boolean") {
    return false;
  }
  if (isError) {
    return typeof errorCode === "string" && errorCode.trim().length > 0;
  }
  return errorCode === undefined;
}
